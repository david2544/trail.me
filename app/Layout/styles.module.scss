@import '/app/Layout/Styles/styles.scss';

.appContainer {
  display: flex;
  height: 100vh;
  flex-direction: column;

  .bodyContainer {
    flex: 1 0 auto;
  }

  .bodyDarkMode {
    transition: background-color 1s ease;
    background-color: $dark-background;
  }

  @media (max-width: 767px) {
    .menuOpen {
      .menuSidebarActive {
        opacity: 1;
      }
    }

    .headerWrapper {
      background-color: $primary;
    }

    .menuSidebarActive {
      background: rgba(0, 0, 0, 0.5);
      position: fixed;
      top: 0;
      left: 0;
      bottom: 0;
      right: 0;
      z-index: 980;
      transition: opacity 200ms ease-in-out;
      opacity: 0;
    }
  }
}
