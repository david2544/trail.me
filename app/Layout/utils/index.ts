import { LINKS, FAIRR_PRODUCTS } from '@utils/constants';

export const isMenuItemActive = {
  isOnProductsPages: (pathname: string) =>
    Object.keys(LINKS.products).some((key: string) => LINKS.products[key].plp === pathname),
  isOnTransferPages: (pathname: string) => Object.values(LINKS.switch).includes(pathname),
  isOnSecurityPage: (pathname: string) => LINKS.security === pathname,
};

export const subHeaderItems = {
  products: [
    {
      link: LINKS.products.riester.plp,
      title: FAIRR_PRODUCTS.riester,
      icon: 'TODO',
      description: 'navi.main.products.riester.description',
    },
    {
      link: LINKS.products.ruerup.plp,
      title: FAIRR_PRODUCTS.ruerup,
      icon: 'TODO',
      description: 'navi.main.products.ruerup.description',
    },
    {
      link: LINKS.products.bav.plp,
      title: FAIRR_PRODUCTS.bav,
      icon: 'TODO',
      description: 'navi.main.products.bav.description',
    },
    {
      link: LINKS.products.robo.plp,
      title: FAIRR_PRODUCTS.robo,
      icon: 'TODO',
      description: 'navi.main.products.robo.description',
    },
    {
      link: LINKS.products.raisin.savings,
      title: 'navi.external.savings-account',
      titleSub: 'navi.external.by-weltsparen',
      description: 'navi.external.savings-account.description',
      isExternal: true,
    },
    {
      link: LINKS.products.raisin.term,
      title: 'navi.external.term-deposit',
      titleSub: 'navi.external.by-weltsparen',
      description: 'navi.external.term-deposit.description',
      isExternal: true,
    },
  ],
  transfer: [
    { link: LINKS.switch.riester, description: 'navi.main.transfer.riester' },
    { link: LINKS.switch.ruerup, description: 'navi.main.transfer.ruerup' },
  ],
};
