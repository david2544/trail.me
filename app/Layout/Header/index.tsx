import React from 'react';
import { Link } from 'react-router-dom';
import { History } from 'history';
import classnames from 'classnames';
import useScroll from '@app/hooks/useScroll';
import DarkModeToggle from 'react-dark-mode-toggle';
import useToggleDarkMode from '@app/hooks/useToggleDarkMode';
import { LINKS } from '@utils/constants';
import Container from '@common/Container';
import Logo from '@images/logo.png';
import { isMenuItemActive } from '../utils';
import styles from './styles.module.scss';

interface HeaderProps {
  history: History;
  toggleMenu: Function;
  menuOpen: boolean;
}

const Header: React.FC<HeaderProps> = ({
  history: {
    location: { pathname },
  },
  toggleMenu,
  menuOpen,
}) => {
  const { isDarkMode, toggleDarkMode } = useToggleDarkMode();

  const { scrollPosition } = useScroll();

  return (
    <div
      className={classnames(styles.mainHeaderNav, {
        [styles.menuOpen]: menuOpen,
        [styles.darkModeHeader]: isDarkMode,
      })}
    >
      <div className={styles.logoContainer}>
        <Container>
          <div
            className={classnames(styles.logoWrapper, {
              [styles.minimized]: scrollPosition > 200,
            })}
          >
            <Link to={LINKS.home}>
              <img alt="fairr-logo" className={styles.logo} src={Logo} />
            </Link>
          </div>
          <button
            type="button"
            tabIndex={0}
            onClick={() => toggleMenu(!menuOpen)}
            className={styles.burgerMenu}
          >
            <span />
            <span />
            <span />
          </button>
        </Container>
      </div>

      <nav className={classnames('container', styles.navContainer)}>
        <ul className={styles.menuWrapper}>
          <li
            className={classnames(styles.menuItemWrapper, {
              [styles.active]: isMenuItemActive.isOnTransferPages(pathname),
            })}
          >
            {/* <Link to={LINKS.switch.general} className={styles.menuItem}>
              navi.main.transfer
            </Link>
            <div className={styles.subNav}>
              <ul>
                {subHeaderItems.transfer.map(item => (
                  <SubNav key={item.description} subMenuItem={item} />
                ))}
              </ul>
            </div> */}
          </li>
          <li
            className={classnames(styles.menuItemWrapper, {
              [styles.active]: isMenuItemActive.isOnSecurityPage(pathname),
            })}
          >
            {/* <Link to={LINKS.security} className={styles.menuItem}>
              about
            </Link> */}
          </li>
          <li className={styles.menuItemWrapper}>
            <div className={`${styles.menuItem} ${styles.darkModeToggle}`}>
              <DarkModeToggle onChange={toggleDarkMode} checked={isDarkMode} size={60} />
            </div>
          </li>
        </ul>
      </nav>
    </div>
  );
};

export default Header;
