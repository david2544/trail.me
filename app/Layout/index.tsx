import React, { useState, useRef } from 'react';
import classnames from 'classnames';
import { useHistory } from 'react-router-dom';
import useOnClickOutside from '@app/hooks/useClickOutside';
import useToggleDarkMode from '@app/hooks/useToggleDarkMode';
import Header from './Header';
import styles from './styles.module.scss';

interface LayoutProps {
  children: React.ReactNode;
}

const Layout: React.FC<LayoutProps> = ({ children }) => {
  const history = useHistory();
  const { isDarkMode } = useToggleDarkMode();
  // Create a ref that we add to the element for which we want to detect outside clicks
  const ref = useRef<HTMLInputElement>(null);
  // State for our modal
  const [menuOpen, toggleMenu] = useState(false);
  // Call hook passing in the ref and a function to call on outside click
  useOnClickOutside(ref, () => toggleMenu(false));

  history.listen(() => {
    if (menuOpen) {
      toggleMenu(false);
    }
  });

  return (
    <div className={styles.appContainer}>
      <div className={classnames({ [styles.menuOpen]: menuOpen })}>
        <div className={classnames({ [styles.headerWrapper]: isDarkMode })} ref={ref}>
          <Header history={history} toggleMenu={toggleMenu} menuOpen={menuOpen} />
        </div>
        {menuOpen && <div className={styles.menuSidebarActive} />}
      </div>
      <div className={classnames(styles.bodyContainer, { [styles.bodyDarkMode]: isDarkMode })}>
        {children}
      </div>
    </div>
  );
};

export default Layout;
