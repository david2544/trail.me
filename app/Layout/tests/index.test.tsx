import React from 'react';
import { shallow } from 'enzyme';
import Layout from '../index';

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as {}),
  useHistory: () => ({
    listen: jest.fn().mockImplementationOnce(cb => {
      cb({ pathname: '/' });
    }),
    location: { pathname: '/' },
  }),
}));

describe('<Layout />', () => {
  const setState = jest.fn();
  const useStateMock: any = () => [true, setState];

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const wrapper = shallow(<Layout>children</Layout>);

    expect(wrapper).toMatchSnapshot();
  });

  it('should toggle menu to false when location changes', () => {
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    shallow(<Layout>children</Layout>);
    expect(setState).toHaveBeenCalledWith(false);
  });
});
