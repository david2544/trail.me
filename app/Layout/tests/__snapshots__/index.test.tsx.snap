// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> should match snapshot 1`] = `
<div
  className="appContainer"
>
  <div
    className=""
  >
    <ClickOutside
      onClickOutside={[Function]}
    >
      <Header
        history={
          Object {
            "listen": [MockFunction] {
              "calls": Array [
                Array [
                  [Function],
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
            "location": Object {
              "pathname": "/",
            },
          }
        }
        menuOpen={false}
        toggleMenu={[Function]}
      />
    </ClickOutside>
  </div>
  <div
    className="bodyContainer"
  >
    children
  </div>
  <Footer
    hasFooterMargin={true}
  />
</div>
`;
